rule all:
    input:
        "results/enhancers_promoters/classified_peaks.bed",
        "results/motif_enrichment/fimo.tsv",
        "results/GREAT_input/great_ready.bed"

rule halLiftover_mapping:
"""
Mapping human liver IDR peaks to mouse genome in mouse coordinates using HALPER as one example.
"""
    input:
    # Unzip and Copy the file of human liver ATAC data from Irene Kaplow /ocean/projects/bio230007p/ikaplow/HumanAtac/Liver/peak/idr_reproducibility/idr.conservative_peak.narrowPeak.gz
    # in your own path the same folder for output
    # If no access to this file, we provide the liver "idr.conservative_peak.narrowPeak" through the link 
    # https://github.com/naithru/713bioinformaticspracticum/blob/main/HALPER_code/idr.conservative_peak.narrowPeak
        "/ocean/projects/bio230007p/jzhao15/project/map_results/Liver_results/HumanLiver_to_MouseLiver/idr.conservative_peak.narrowPeak"
    output:
        # Need to edit in your path
        # Outputs generated by HALPER (3 files)
        bed_sfile = "/ocean/projects/bio230007p/jzhao15/project/map_results/Liver_results/HumanLiver_to_MouseLiver/idr.conservative_peak.HumanToMouse.halLiftover.sFile.bed",
        bed_tfile = "/ocean/projects/bio230007p/jzhao15/project/map_results/Liver_results/HumanLiver_to_MouseLiver/idr.conservative_peak.HumanToMouse.halLiftover.tFile.bed",
        mapped_peaks = "/ocean/projects/bio230007p/jzhao15/project/map_results/Liver_results/HumanLiver_to_MouseLiver/idr.conservative_peak.HumanToMouse.HALPER.narrowPeak"
    
   params:
        # Download the script from the link https://github.com/naithru/713bioinformaticspracticum/blob/main/HALPER_code/liver_map.job in your own path
        job_script = "/ocean/projects/bio230007p/jzhao15/project/HALPER_code/liver_map.job"
        
    shell:
        """
        bash {params.job_script}
        """
rule overlap_analysis_liver:
    """
    Intersect HALPER-mapped human liver peaks with mouse liver ATAC-seq peaks.
    Calculates how many human regions are still open in the mouse liver.
    
    NOTE: Update the paths below to match your own directory structure.
    """
    #Input is a result from HALPER
    input:
        halper_mapped = "/ocean/projects/bio230007p/rammohan/project/map_results/Liver_results/HumanLiver_to_MouseLiver/idr.conservative_peak.HumanToMouse.HALPER.narrowPeak",
        mouse_peaks = "/ocean/projects/bio230007p/rammohan/project/map_results/Liver_results/MouseLiver_peaks.bed"
    output:
        overlap = "/ocean/projects/bio230007p/rammohan/project/map_results/Liver_results/HumanLiver_to_MouseLiver/halper_liver_overlap.bed"
    params:
        # Download the script from the link https://github.com/naithru/713bioinformaticspracticum/blob/main/overlap_analysis/overlap_liver.job in your own path
        job_script = "/ocean/projects/bio230007p/rammohan/project/map_results/Liver_results/HumanLiver_to_MouseLiver/overlap_liver.job"
    shell:
        """
        bash {params.job_script}
        """
        
rule classify_elements:
    input:
        "results/overlap/overlap_peaks.bed"
    output:
        "results/enhancers_promoters/classified_peaks.bed"
    shell:
        "bash scripts/classify_elements.sh {input} {output}"

rule motif_enrichment:
    input:
        "results/enhancers_promoters/classified_peaks.bed"
    output:
        "results/motif_enrichment/fimo.tsv"
    shell:
        "bash scripts/run_motif_analysis.sh {input} {output}"

rule prepare_GREAT_input:
    input:
        "results/enhancers_promoters/classified_peaks.bed"
    output:
        "results/GREAT_input/great_ready.bed"
    shell:
        "bash scripts/prepare_great_input.sh {input} {output}"

